### ISSUE TITLE
Cereal incorrectly infers argument type for "units" objects

### ISSUE THREAD

**Issue by osadowskyphd** [USCiLab/cereal#484]
I am trying to create templated code to serialize types from the "units" library -- available at https://github.com/nholthaus/units.  The specific library is probably not the critical issue, but it is a useful example.  The units library defines traits which allow the programmer to test if a certain type belongs to units or not, e.g. by using std::enable_if.

I use Cereal 1.2.2, units 2.3.0, g++ 7.2 on Ubuntu 16

Here is the templated code I wrote.
//==================================================
#include <cstdio>
#include <cstdint>
#include <inttypes.h>
#include "units.h"
#include <cereal/archives/json.hpp>

using namespace units;
using namespace units::angle;
using namespace units::literals;

namespace cereal
{
    template<class Archive, class _serializedType>
    typename std::enable_if<::units::traits::is_unit_t<_serializedType>::value, typename _serializedType::underlying_type>::type
    save_minimal(Archive &archive, _serializedType const &unitData)
    {
        return unitData.value();
    }

    template<class Archive, class _serializedType>
    typename std::enable_if<::units::traits::is_unit_t<_serializedType>::value, void>::type
    load_minimal(Archive &archive, _serializedType &unitData, typename _serializedType::underlying_type const &value)
    {
        unitData = _serializedType(value);
    }
}
int main()
{
    units::angle::degree_t geocoord_deg = 48.3264284_deg;
    std::printf("geocoord_deg %14.10f\n", geocoord_deg.value());

    std::stringstream outputString;
    {
        cereal::JSONOutputArchive archive(outputString);
        archive(CEREAL_NVP(geocoord_deg));
    }
    std::cout << "Written as JSON string:\n" << outputString.str() << std::endl;

    {
        std::stringstream inputString(outputString.str());
        cereal::JSONInputArchive archive(inputString);
        //cereal::load_minimal(archive, geocoord_deg, 0.0);
        archive(geocoord_deg);
    }
        std::cout << "Read from JSON: " << geocoord_deg() << std::endl;
    return 0;
}
//=====================================================
The code generates a compiler error ending with the following messages:
/media/psf/Home/SoftwarePackages/cereal-1.2.2/include/cereal/details/traits.hpp:862:5: error: static assertion failed: cereal detected an invalid serialization type parameter in non-member load_minimal.  load_minimal non-member functions must accept their serialization type by non-const reference
     CEREAL_MAKE_HAS_NON_MEMBER_LOAD_MINIMAL_TEST(load_minimal, save_minimal, )

Notice that the load_minimal function that I defined does not have a const _serializedType reference, although _serializedType could somehow be inferred to be const.

Interestingly, if I substitute the template argument _serializedType with the concrete type units::angle::degree_t, the compilation is successful.

namespace cereal
{
    template<class Archive>
    std::enable_if<::units::traits::is_unit_t<::units::angle::degree_t>::value, ::units::angle::degree_t::underlying_type>::type
    save_minimal(Archive &archive, ::units::angle::degree_t const &unitData)
    {
        return unitData.value();
    }

    template<class Archive>
    std::enable_if<::units::traits::is_unit_t<::units::angle::degree_t>::value, void>::type
    load_minimal(Archive &archive, ::units::angle::degree_t &unitData, ::units::angle::degree_t::underlying_type const &value)
    {
        unitData = ::units::angle::degree_t(value);
    }
}

Any idea what could have gone wrong in the first version?

Regards

Ofri Sadowsky

**Comment by AzothAmmo** [comment]
I think this is the same problem as #132, which is to say that our type checks for minimal serialization don't seem to play nicely with partial template specialization. It's actually a very hard problem to find a good fix for. The best advice I can give, if this is indeed the same issue, is to comment out the assertion that fails as described in #132.

**Comment by osadowskyphd** [comment]
Yes, this looks very similar.  I am not planning to change anything in the library, though.  I'll have a workaround like a macro.
